### DISCUSSION TITLE
How to set a timestampz value in the DB?

### DISCUSSION THREAD

**Discussion by farzd** [supabase/supabase#2625] [Questions] [answered]
Postgres is set to UTC,
column type is 'timestampz'
I'm interested in 'Europe/Dublin'

I have a lambda running on AWS Dublin so my query is, which i think is fine?
```
  const today = new Date();
  let tomorrow = new Date();
  tomorrow.setDate(today.getDate() + 1);

  let { data: quizzes, error: quizzesErr } = await supabase
    .from('quizzes')
    .select('*')
    .gte('startDate', today.toISOString())
    .lte('endDate', tomorrow.toISOString());
```

But when i set a date, its one hour behind, the way i'm setting the date in DB is
i'm using https://github.com/marnusw/date-fns-tz
```
const timeZone = 'Europe/Dublin';
zonedTimeToUtc(dates.startDate, timeZone).toISOString(),
```
I noticed that its only correct when i use date picker in supabase/grid
so had a look at https://github.com/supabase/grid/blob/314fe946c0185b124088e8490906ee06f72777cb/src/components/editor/DateTimeEditor.tsx
and seems like its just formatting the date, using day.js
```
const SERVER_DATE_TIME_FORMAT = 'YYYY-MM-DDTHH:mm:ssZ';
// dayjs(value, SERVER_DATE_TIME_FORMAT).format(INPUT_DATE_TIME_FORMAT)
```

i tried to do this with date-fns-tz library but still getting inconsistent results

![image](https://user-images.githubusercontent.com/1423413/127502770-6553869c-e6bc-4f34-9308-bc014952f206.png)
[i expect the date to look like the first row, with the offset in there, which seems to only be possible for me when i use the date picker in the supabase/grid]

**Comment by dshukertjr** [comment, marked as answer]
Hi @farzd!

It sounds like you want to store the timestamp values in your Supabase database using `timestampz` type, but having trouble, correct?

When you store  a `timestampz` values in your database, the timestamp and timezone values are auto adjusted as soon as they are saved. Timezone is adjusted to the timezone set on your database, which is `UTC` by default for Supabase. 

In your case, when you store for example `2021-07-28T17:00:00+01:00`, on the server it will be stored as `UTC` like this `2021-07-28T16:00:00+00:00`. If you refresh your page after you insert a `timestampz` value into you Supabase database from the console, you should see that the timezone value gets set to 00:00 .

This is usually fine, because when you use it on the client side, you will be converting to `Date` objects, which will take timezones into account. For example in JavaScript, you can do this:
```javascript
const date = new Date('2021-07-28T16:00:00+00:00')
// date will be in whatever the browser's timezone setting is. 
```

I hope this answer helps you, but if there's anything I can further assist, please let me know! Also, you can read more about how `timezonez` behaves in Postgres [here](https://www.postgresqltutorial.com/postgresql-timestamp/).

**Reply by fergusmeiklejohn** [reply]
Sometimes we need to show what the time will be in another timezone (like if you are running a conference in Beijing, you'll want to show the time of a talk in Beijing not only in the browser of the user wherever they might be. So then we can save the timezone along with the timestamp:
```
    "scheduled": "2021-10-26T16:00:00.000+08:00",
    "timezone": "UTC+8",
 ```
So then we can show "9:30am - Supabase Keynote" or "1:30am - Supabase Keynote" depending on what the user wants

**Reply by dshukertjr** [reply]
Hi @fergusmeiklejohn!

Since `timestampz` already has the timezone information, you don't really need to store the timezone as a separate column. You can do something like [this](https://stackoverflow.com/questions/10087819/convert-date-to-another-timezone-in-javascript) to change the timezone of the displaying time on the client side if browser's timezone is not the what you want to show!

**Reply by farzd** [reply]
Thanks for the response @dshukertjr , i appreciate thats how it works. However there are some strange inconsistencies  when playing with the HTML input datetime-local. This is how i got it to work, incase anyone comes across this in the future

reading the UTC date and rendering it on datetime-local HTML form 
```
import { format, utcToZonedTime } from 'date-fns-tz';

const formaStr = `yyyy-MM-dd'T'HH:mm:ss`;
const timeZone = 'Europe/Dublin';
const startDate = format(utcToZonedTime(result.startDate, timeZone), formaStr, {timeZone})
```

saving the date to DB in UTC
```
new Date(startDate).toISOString(),
```

and querying 
```
....lte('startDate', today.toISOString())
```

**Reply by fergusmeiklejohn** [reply]
@dshukertjr I'm confused because in my experience saving a timezone date to postgres and retrieving it, I lose the original timezone:
I save this to postgres:
```json
[
  {
    "scheduled": "2021-10-12T14:00:00.000-04:00"
  }
]
```

And when I pull it from the database the time is UTC correct but without information about what the original timezone was:
```json
[
  {
    "scheduled": "2021-10-12 18:00:00.000000 +00:00"
  }
]
```
